File: src/core/Market/market_manager.py
import logging
import time
import threading
from threading import Lock
from typing import Dict, List, Optional
from src.core.Market.item_type import ItemType
from src.core.Market.market_item import MarketItem
from src.core.Market.item_stock import ItemStock


class MarketManager:
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        self._lock = Lock()
        self.active_items: Dict[str, MarketItem] = {}
        self.seller_stocks: Dict[str, Dict[ItemType, ItemStock]] = {}
        self.logger.info("Initializing market manager")
        self.initialize_market_items()

    def initialize_market_items(self):
        """Initialize the market with basic items"""
        with self._lock:
            try:
                for item_type in ItemType:
                    item_id = f"market_{item_type.value}"
                    item = MarketItem(
                        item_id=item_id,
                        item_type=item_type,
                        quantity=5.0,
                        seller_id="market",
                        max_sale_duration=60,
                    )
                    self.active_items[item_id] = item
                    self.logger.info(f"Initialized market item: {item.to_dict()}")
            except Exception as e:
                self.logger.error(f"Error initializing market items: {e}")
                raise

    def initialize_seller_stock(self, seller_id: str, initial_quantity: float = 5.0):
        """Initialize a seller's stock"""
        with self._lock:
            if seller_id in self.seller_stocks:
                return  # Already initialized

            self.seller_stocks[seller_id] = {}
            for item_type in ItemType:
                self.seller_stocks[seller_id][item_type] = ItemStock(
                    item_type=item_type,
                    quantity=initial_quantity,
                    max_quantity=initial_quantity,
                )
            self.logger.info(f"Initialized stock for seller {seller_id}")

    def get_active_items(self) -> List[dict]:
        """Get list of all active items"""
        with self._lock:
            items = []
            for item in self.active_items.values():
                if not item.is_expired():
                    try:
                        items.append(item.to_dict())
                    except Exception as e:
                        self.logger.error(f"Error serializing item: {e}")
            return items

    def try_purchase(self, item_id: str, quantity: float) -> bool:
        """Attempt to purchase an item"""
        with self._lock:
            if item_id not in self.active_items:
                raise ValueError(f"Item {item_id} not found")

            item = self.active_items[item_id]
            if item.is_expired():
                raise ValueError("Item sale has expired")

            return item.try_purchase(quantity)

    def start_sale(
        self, seller_id: str, item_type: ItemType, quantity: float
    ) -> MarketItem:
        """Start a new sale"""
        with self._lock:
            # Validate seller and check stock
            if seller_id not in self.seller_stocks:
                raise ValueError(f"Seller {seller_id} not found")

            seller_stock = self.seller_stocks[seller_id]
            if item_type not in seller_stock:
                raise ValueError(f"Seller does not have {item_type.value} in stock")

            stock = seller_stock[item_type]
            if stock.quantity < quantity:
                raise ValueError(f"Insufficient stock for {item_type.value}")

            # Create new sale
            item_id = f"sale_{seller_id}_{int(time.time())}"
            item = MarketItem(
                item_id=item_id,
                item_type=item_type,
                quantity=quantity,
                seller_id=seller_id,
            )

            # Update stock and add to active items
            stock.quantity -= quantity
            self.active_items[item_id] = item

            self.logger.info(f"Started sale: {item.to_dict()}")
            return item

    def end_sale(self, item_id: str) -> None:
        """End a sale"""
        with self._lock:
            if item_id not in self.active_items:
                return

            item = self.active_items[item_id]
            if item.quantity > 0 and item.seller_id in self.seller_stocks:
                # Return unsold quantity to stock
                seller_stock = self.seller_stocks[item.seller_id]
                if item.item_type in seller_stock:
                    stock = seller_stock[item.item_type]
                    stock.quantity += item.quantity

            del self.active_items[item_id]
            self.logger.info(f"Ended sale: {item_id}")

    def get_seller_stock(self, seller_id: str) -> Dict[str, float]:
        """Get current stock levels for a seller"""
        with self._lock:
            if seller_id not in self.seller_stocks:
                raise ValueError(f"Seller {seller_id} not found")

            return {
                item_type.value: stock.quantity
                for item_type, stock in self.seller_stocks[seller_id].items()
            }

    def handle_buy_request(self, item_id: str, quantity: float, buyer_id: str) -> bool:
        """Handle a buy request"""
        with self._lock:
            try:
                if item_id not in self.active_items:
                    self.logger.warning(f"Item {item_id} not found for purchase")
                    return False

                result = self.try_purchase(item_id, quantity)
                if result:
                    self.logger.info(
                        f"Buyer {buyer_id} purchased {quantity} of {item_id}"
                    )
                return result
            except Exception as e:
                self.logger.error(f"Error processing buy request: {e}")
                return False

    def get_market_status(self) -> dict:
        """Get current market status"""
        with self._lock:
            active_items = self.get_active_items()
            return {
                "active_items": len(active_items),
                "items": active_items,
                "sellers": len(self.seller_stocks),
                "total_quantity": sum(item["quantity"] for item in active_items),
            }

    def cleanup_expired_items(self) -> None:
        """Remove expired items"""
        with self._lock:
            expired = [
                item_id
                for item_id, item in self.active_items.items()
                if item.is_expired()
            ]
            for item_id in expired:
                self.end_sale(item_id)

File: app/src/main/java/main/java/main/market/MarketManager.java
package main.java.main.market;

import java.util.concurrent.*;
import java.util.*;
import java.util.logging.*;
import java.util.stream.Collectors;

public class MarketManager {
    private static final Logger logger = Logger.getLogger(MarketManager.class.getName());
    private final ConcurrentHashMap<String, Item> activeItems = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<String, Map<String, Double>> sellerStocks = new ConcurrentHashMap<>();
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();

    public MarketManager() {
        scheduler.scheduleAtFixedRate(this::cleanupExpiredItems, 1, 1, TimeUnit.SECONDS);
    }

    public void initializeSellerStock(String sellerId) {
        Map<String, Double> stock = new ConcurrentHashMap<>();
        stock.put("flower", 1000.0);
        stock.put("sugar", 1000.0);
        stock.put("potato", 1000.0);
        stock.put("oil", 1000.0);
        sellerStocks.put(sellerId, stock);
        logger.info("Initialized stock for seller: " + sellerId);
    }

    public synchronized Item startSale(String sellerId, String itemName, double quantity) {
        Map<String, Double> stock = sellerStocks.get(sellerId);
        if (stock == null) {
            throw new IllegalStateException("Seller not found: " + sellerId);
        }

        Double available = stock.get(itemName);
        if (available == null || available < quantity) {
            throw new IllegalStateException("Insufficient stock for " + itemName);
        }

        // Update stock
        stock.put(itemName, available - quantity);

        // Create new item
        String itemId = "sale_" + sellerId + "_" + System.currentTimeMillis();
        // Pass the duration in seconds (60 is the max allowed per requirements)
        Item item = new Item(itemId, itemName, quantity, sellerId, 60);
        activeItems.put(itemId, item);

        logger.info(String.format("Sale started: %s, quantity: %.2f, seller: %s", 
                    itemName, quantity, sellerId));
        return item;
    }

    public synchronized boolean handleBuyRequest(String itemId, double quantity, String buyerId) {
        Item item = activeItems.get(itemId);
        if (item == null) {
            logger.warning("Item not found: " + itemId);
            return false;
        }

        if (item.isExpired()) {
            logger.warning("Item expired: " + itemId);
            return false;
        }

        boolean success = item.tryPurchase(quantity);
        if (success) {
            logger.info(String.format("Purchase successful: %.2f of %s by %s", 
                        quantity, itemId, buyerId));
        } else {
            logger.warning(String.format("Purchase failed: %.2f of %s by %s", 
                        quantity, itemId, buyerId));
        }
        return success;
    }

    public synchronized void endSale(String sellerId) {
        List<Item> sellerItems = activeItems.values().stream()
                .filter(item -> item.getSellerId().equals(sellerId))
                .collect(Collectors.toList());

        if (sellerItems.isEmpty()) {
            logger.info("No active sales found for seller: " + sellerId);
            return;
        }

        for (Item item : sellerItems) {
            endSingleSale(item.getId());
        }
        logger.info("Ended all sales for seller: " + sellerId);
    }

    private synchronized void endSingleSale(String itemId) {
    Item item = activeItems.get(itemId);
    if (item != null) {
        item.forceClose(); // Force close the item
        activeItems.remove(itemId);
        double remainingQuantity = item.getQuantity();
        if (remainingQuantity > 0) {
            Map<String, Double> stock = sellerStocks.get(item.getSellerId());
            if (stock != null) {
                stock.merge(item.getName(), remainingQuantity, Double::sum);
                logger.info(String.format("Sale ended: Returned %.2f %s to seller %s stock", 
                    remainingQuantity, item.getName(), item.getSellerId()));
            }
        }
    }
}
   

    public List<Item> getActiveItems() {
        return activeItems.values().stream()
                .filter(item -> !item.isExpired())
                .collect(Collectors.toList());
    }

    public String getSellerIdForItem(String itemId) {
        Item item = activeItems.get(itemId);
        return item != null ? item.getSellerId() : null;
    }

    private void cleanupExpiredItems() {
        List<Item> expiredItems = activeItems.values().stream()
                .filter(Item::isExpired)
                .collect(Collectors.toList());

        for (Item item : expiredItems) {
            endSingleSale(item.getId());
            logger.info("Cleaned up expired item: " + item.getId() + 
                       " (Remaining time: " + item.getRemainingTime() + "ms)");
        }
    }

    public void shutdown() {
        scheduler.shutdown();
        try {
            if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                scheduler.shutdownNow();
            }
        } catch (InterruptedException e) {
            scheduler.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
}